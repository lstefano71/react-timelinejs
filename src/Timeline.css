/* axis */

/*.axis { !* axis labels *!*/
    /*fill: #808080;*/
    /*font-family: sans-serif;*/
    /*font-size: 10px;*/
/*}*/

/*.axis line{ !* axis tick marks *!*/
    /*stroke-width : 1;*/
    /*stroke: grey;*/
    /*shape-rendering: crispEdges;*/
/*}*/

/*.axis path { !* axis line *!*/
    /*stroke-width : 1;*/
    /*stroke: grey;*/
    /*shape-rendering: crispEdges;*/
/*}*/

/* timeline band */

.timeline {
    position: relative;
}

.band { /* band background */
    fill: #FAFAFA;
}

/* labels */

.bandLabel {
    fill: #F0F0F0;
    font: 10px sans-serif;
    font-weight: bold;
}

.bandMinMaxLabel {
    fill: blue;
    font: 10px sans-serif;
    font-weight: bold;
}

.bandMidLabel {
    fill: red;
    font: 10px sans-serif;
    font-style: italic;
    font-weight: bold;
}

/* brush */

.brush .extent {
    stroke: gray;
    fill: blue;
    fill-opacity: .1;
}

.chart {
    fill: #EEEEEE;
}

.interval {
    fill: dodgerblue;
    stroke-width: 6;
    cursor : default;
    pointer-events: true;
}

.instant {
    fill: dodgerblue;
    stroke-width: 6;
    cursor : default;
    /*pointer-events: true;*/
}

.instantLabel {
    fill : blue;
    font: 10px sans-serif;
    shape-rendering: crispEdges;
}

.intervalLabel {
    fill : black;
    font: 10px sans-serif;
    shape-rendering: crispEdges;
}

.item {
    cursor : default;
    pointer-events: auto;
}

/*.svg {*/
    /*border-style: solid;*/
    /*border-width: 1px;*/
    /*border-color: black;*/
    /*background-color: #FFFFFF;*/
/*}*/

.tooltip {
    width: 0px;
    height: 0px;
    position: absolute;
    z-index: 2;
    visibility: hidden;
    color : black;
    cursor:default;
    pointer-events: none;
    display: flex;
}

.tooltip.top-left {

}
.tooltip.top-right {
    flex-direction: row-reverse;
}
.tooltip.bottom-right {
    flex-flow: row-reverse wrap-reverse;
}
.tooltip.bottom-left {
    flex-flow: column-reverse;
}

.tooltip-container {
    shape-rendering: crispEdges;
    padding: 4px;
    min-width: 300px;
    min-height: 100px;
    border: 1px solid;
    background-color: #FFFFEE;
}

.mouseover-line {
    stroke: #0984e3;
}

.mouseover-label {
    fill : white;
    font: 10px sans-serif;
    shape-rendering: crispEdges;
    text-anchor: middle;
}

.mouseover-label-background {
    fill: #0984e3;
}
