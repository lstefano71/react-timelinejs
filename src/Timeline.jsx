import React from 'react';
import timeline from './TimelineModel';
import './Timeline.css';
import equals from 'ramda/src/equals';
import PropTypes from 'prop-types';

const propTypes = {
    data: PropTypes.array.isRequired,
    lines: PropTypes.array,

    dataKey: PropTypes.string,
    width: PropTypes.number,
    height: PropTypes.number,
    trackHeight: PropTypes.number,
    label: PropTypes.bool,
    tooltips: PropTypes.bool,
    tooltipContent: PropTypes.func,
    brush: PropTypes.bool,
    brushRange: PropTypes.array,
    range: PropTypes.array,
    onBrush: PropTypes.func,
    onMouseover: PropTypes.func,
    onClick: PropTypes.func,
};

//
// const defaultProps = {
//     styles: {
//         label: {
//             fontFamily: 'Comic Sans MS',
//             color: 'green'
//         },
//         input: {
//             background: '#ddd',
//             border: '1px solid red'
//         }
//     }
// };

class Timeline extends React.Component {

    componentDidMount() {
        this.key = Math.round(Math.random() * 100000).toFixed(0) + '-' + new Date().getTime();
        this.createTimeline();
    }


    componentDidUpdate() {
    }

    componentWillReceiveProps(nextProps, nextState) {
        if (!equals(nextProps.data, this.props.data)
                || !equals(nextProps.lines, this.props.lines)) {
            this.timelineFn
                .data(nextProps.data)
                .lines(nextProps.lines)
                .update()
                .updateRange(nextProps.range)
                .updateBrushRange(nextProps.brushRange)
                .redraw();
        }
        if (!equals(nextProps.range, this.props.range)){
            this.timelineFn.updateRange(nextProps.range);
        }
        if (!equals(nextProps.brushRange, this.props.brushRange)){
            this.timelineFn.updateBrushRange(nextProps.brushRange);
        }
    }

    createTimeline = () => {

        const {data, dataKey, lines, height, width, trackHeight, label, tooltips, tooltipContent, brush, brushRange, onBrush, onMouseover, onClick, range} = this.props;
        const config = {
            dataKey,
            width,
            height,
            trackHeight,
            label,
            tooltips,
            tooltipContent,
            brush,
            brushRange,
            range,
            onBrush,
            onMouseover,
            onClick
        };
        this.timelineFn = timeline(this.div, config).create(data, lines);
        this.timelineFn.redraw();
    };

    render() {
        return <div ref={input => this.div = input}/>
    }
}

Timeline.propTypes = propTypes;
// Timeline.defaultProps = defaultProps;

export default Timeline;
