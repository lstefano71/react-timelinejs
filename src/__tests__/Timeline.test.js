import React from 'react';
import renderer from 'react-test-renderer';
import Timeline from '../Timeline';
import {mockData} from './Timeline.tester';

test('render empty timeline', () => {

    const timeline = mount(<Timeline data={[]}/>);
    expect(timeline).toMatchSnapshot();

});

test('render with sample data', () => {

    const timeline = mount(<Timeline data={mockData({size: 3})}/>);
    expect(timeline).toMatchSnapshot();

});