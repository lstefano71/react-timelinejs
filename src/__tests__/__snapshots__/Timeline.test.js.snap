// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render empty timeline 1`] = `
<Timeline
  data={Array []}
>
  <div />
</Timeline>
`;

exports[`render with sample data 1`] = `
<Timeline
  data={
    Array [
      Object {
        "end": 2018-07-14T22:15:56.060Z,
        "instant": false,
        "label": "event 2",
        "start": 2018-07-14T21:01:56.060Z,
        "steps": Array [
          Object {
            "className": "custom_1",
            "end": 2018-07-14T21:24:08.060Z,
            "endPercentage": 0.3,
            "label": "sub event 1",
            "start": 2018-07-14T21:09:20.060Z,
            "startPercentage": 0.1,
          },
          Object {
            "className": "custom_1",
            "end": 2018-07-14T21:38:56.060Z,
            "endPercentage": 0.5,
            "label": "sub event 2",
            "start": 2018-07-14T21:31:32.060Z,
            "startPercentage": 0.4,
          },
          Object {
            "className": "custom_1",
            "end": 2018-07-14T21:57:26.060Z,
            "endPercentage": 0.75,
            "label": "sub event 3",
            "start": 2018-07-14T21:42:38.060Z,
            "startPercentage": 0.55,
          },
        ],
        "track": 0,
      },
      Object {
        "end": 2018-07-14T02:28:37.781Z,
        "instant": false,
        "label": "event 0",
        "start": 2018-07-14T01:42:37.781Z,
        "steps": Array [
          Object {
            "className": "custom_1",
            "end": 2018-07-14T01:56:25.781Z,
            "endPercentage": 0.3,
            "label": "sub event 1",
            "start": 2018-07-14T01:47:13.781Z,
            "startPercentage": 0.1,
          },
          Object {
            "className": "custom_1",
            "end": 2018-07-14T02:05:37.781Z,
            "endPercentage": 0.5,
            "label": "sub event 2",
            "start": 2018-07-14T02:01:01.781Z,
            "startPercentage": 0.4,
          },
          Object {
            "className": "custom_1",
            "end": 2018-07-14T02:17:07.781Z,
            "endPercentage": 0.75,
            "label": "sub event 3",
            "start": 2018-07-14T02:07:55.781Z,
            "startPercentage": 0.55,
          },
        ],
        "track": 0,
      },
      Object {
        "end": 2018-07-14T00:52:46.043Z,
        "instant": false,
        "label": "event 1",
        "start": 2018-07-14T00:17:46.043Z,
        "steps": Array [
          Object {
            "className": "custom_1",
            "end": 2018-07-14T00:28:16.043Z,
            "endPercentage": 0.3,
            "label": "sub event 1",
            "start": 2018-07-14T00:21:16.043Z,
            "startPercentage": 0.1,
          },
          Object {
            "className": "custom_1",
            "end": 2018-07-14T00:35:16.043Z,
            "endPercentage": 0.5,
            "label": "sub event 2",
            "start": 2018-07-14T00:31:46.043Z,
            "startPercentage": 0.4,
          },
          Object {
            "className": "custom_1",
            "end": 2018-07-14T00:44:01.043Z,
            "endPercentage": 0.75,
            "label": "sub event 3",
            "start": 2018-07-14T00:37:01.043Z,
            "startPercentage": 0.55,
          },
        ],
        "track": 0,
      },
    ]
  }
>
  <div />
</Timeline>
`;
